src/components/Microsoft.DotNet.UpgradeAssistant/TargetFramework/TargetFrameworkSelector.cs

```diff
@@ -0,0 +1,104 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.DotNet.UpgradeAssistant.TargetFramework
+{
+    public class TargetFrameworkSelector : ITargetFrameworkSelector
+    {
+        private const string DefaultCurrentTFMBase = "net5.0";
+        private const string DefaultLTSTFMBase = "net5.0";
+
+        private readonly ITargetFrameworkMonikerComparer _comparer;
+        private readonly IEnumerable<ITargetFrameworkSelectorFilter> _selectors;
+
+        private readonly string _currentTFMBase;
+        private readonly string _ltsTFMBase;
+
+        private readonly UpgradeTarget _upgradeTarget;
+
+        public TargetFrameworkSelector(UpgradeOptions options, ITargetFrameworkMonikerComparer comparer, IOptions<TFMSelectorOptions> selectorOptions, IEnumerable<ITargetFrameworkSelectorFilter> selectors)
+        {
+            _comparer = comparer;
+            _selectors = selectors;
+
+            _currentTFMBase = selectorOptions?.Value.CurrentTFMBase ?? DefaultCurrentTFMBase;
+            _ltsTFMBase = selectorOptions?.Value.LTSTFMBase ?? DefaultLTSTFMBase;
+            _upgradeTarget = options?.UpgradeTarget ?? throw new ArgumentNullException(nameof(options));
+        }
+
+        public async ValueTask<TargetFrameworkMoniker> SelectTargetFrameworkAsync(IProject project, CancellationToken token)
+        {
+            if (project is null)
+            {
+                throw new ArgumentNullException(nameof(project));
+            }
+
+            var appBase = _upgradeTarget == UpgradeTarget.Current ? _currentTFMBase : _ltsTFMBase;
+            var current = GetNetStandardTFM(project);
+            var appBaseTfm = new TargetFrameworkMoniker(appBase);
+
+            var updater = new FilterState(_comparer, project, current, appBaseTfm)
+            {
+                Components = await project.GetComponentsAsync(token).ConfigureAwait(false),
+            };
+
+            foreach (var selector in _selectors)
+            {
+                selector.Process(updater);
```

Comment:
I see now that this is more of the decorator pattern with conditionals. https://www.dofactory.com/net/decorator-design-pattern

We may want to order the decorators

Reply:
There's no ordering involved here - we ensure that you can never downgrade a TFM in a filter.

Reply:
Yeah, that makes sense. I still think it could apply though.

One thing the order could do is put the SatisfiesProjectDependencies step at last and skip a bunch of NuGet calls if you're already targetting net5.0?

Reply:
What NuGet calls are you seeing? It's just looking at the TFM the dependencies are currently targeting.

Reply:
Sorry, I was still using my old understanding of `SatisifiesProjectDependenciesTargetFrameworkSelectorFilter` which does not make NuGet calls.

Reply:
Does this not have the opportunity to `break` the loop if the tfm is already as net5.0?

Reply:
We don't want to break the loop because there's no way to know what future selectors might do. Even if we're targeting net5.0, it's possible that a future selector (min dependency selector, for example) could require a TFM higher than our app base!